doc/sphinx/source/users-guide/locality/onClauses.rst

```diff
@@ -0,0 +1,111 @@
+  :language: text
+
+Here's a minor change to this example which demonstrates that the
+original task starts and ends on locale #0:
+
+.. literalinclude:: ../../../../../test/release/examples/users-guide/locality/onClausePlusOrig.chpl
+  :language: chapel
+
+When run on three locales, this program generates:
+
+.. literalinclude:: ../../../../../test/release/examples/users-guide/locality/onClausePlusOrig.good
+  :language: text
+  :lines: 1-12
+
+Note that the statement following an on-clause may include compound
+statements or function calls, so can result in an arbitrary amount of
+code being executed remotely.
+
+
+Locality is Orthogonal to Parallelism
+-------------------------------------
+
+It's important to note that Chapel's features for controlling locality
+are orthogonal to those used to create parallelism.  Specifically,
+while both of the above programs run using multiple locales, neither
+is a parallel program.  Instead, a single conceptual task is moving
+between locales as it executes.  Similarly, all parallel programs from
+previous sections will only execute on a single locale---that is, they
+only result in shared-memory parallelism.
+
+In order to get distributed memory parallelism, all that's required is
+to compose Chapel's features for parallelism and locality.  As a
+simple example, the following program uses a coforall loop to create a
+task per locale in combination with an on-clause to execute each task
+on its corresponding locale:
+
+.. literalinclude:: ../../../../../test/release/examples/users-guide/locality/coforallPlusOn.chpl
+  :language: chapel
+
+In effect, this creates an SPMD (Single Program, Multiple Data) style
+of parallelism.  When running on four locales, the output will be a
+nondeterministic permutation of the following four lines:
+
+.. literalinclude:: ../../../../../test/release/examples/users-guide/locality/coforallPlusOn.good
+  :language: text
+
+
+Data-driven On-clauses
+----------------------
+
+While the on-clauses in the examples above use a locale value to
+specify where execution should take place, more generally, on-clauses
+can be specified using arbitrary expressions (typically variable
+references).  In such cases, execution moves to the locale on which
+the variable or expression is stored.  These are referred to as
+*data-driven on-clauses*.
+
+As a simple example of a data-driven on-clause, the following program
+moves its task to the locale on which *x* is stored:
+
+.. literalinclude:: ../../../../../test/release/examples/users-guide/locality/dataDrivenOnClause-trivial.chpl
+  :language: chapel
+
+However, this program is trivial in that *x* is stored on locale 0,
+so the on-clause is essentially degenerate and doesn't move execution
+anywhere:
+
+.. literalinclude:: ../../../../../test/release/examples/users-guide/locality/dataDrivenOnClause-trivial.good
+  :language: text
+
+In the next section, we'll see more about how variables can be
```

Comment:
The "more about" here could be removed, since we haven't seen a first instance of declaring variables on other locales yet.